crates/mdk-core/src/image_processing/validation.rs

```diff
@@ -0,0 +1,501 @@
+        })?;
+
+    let format = img_reader
+        .format()
+        .ok_or_else(|| ImageProcessingError::InvalidMimeType {
+            mime_type: "Could not determine image format".to_string(),
+        })?;
+
+    // Convert image::ImageFormat to MIME type
+    let mime_type = match format {
+        image::ImageFormat::Png => "image/png",
+        image::ImageFormat::Jpeg => "image/jpeg",
+        image::ImageFormat::Gif => "image/gif",
+        image::ImageFormat::WebP => "image/webp",
+        image::ImageFormat::Bmp => "image/bmp",
+        image::ImageFormat::Ico => "image/x-icon",
+        image::ImageFormat::Tiff => "image/tiff",
+        image::ImageFormat::Tga => "image/x-tga",
+        image::ImageFormat::Dds => "image/vnd-ms.dds",
+        image::ImageFormat::Hdr => "image/vnd.radiance",
+        image::ImageFormat::OpenExr => "image/x-exr",
+        image::ImageFormat::Pnm => "image/x-portable-anymap",
+        image::ImageFormat::Farbfeld => "image/x-farbfeld",
+        image::ImageFormat::Avif => "image/avif",
+        image::ImageFormat::Qoi => "image/qoi",
+        _ => {
+            return Err(ImageProcessingError::InvalidMimeType {
+                mime_type: format!("Unsupported image format: {:?}", format),
+            });
+        }
+    };
+
+    Ok(mime_type.to_string())
+}
+
+/// Validate that the provided MIME type matches the actual file data
+///
+/// This function detects the actual image format from the file data and compares
+/// it with the provided MIME type. This prevents MIME type confusion attacks where
+/// an attacker provides a misleading MIME type.
+///
+/// # Arguments
+/// * `data` - The image file data
+/// * `claimed_mime_type` - The MIME type claimed by the application
+///
+/// # Returns
+/// * The canonical MIME type (validated and normalized)
+///
+/// # Errors
+/// * `InvalidMimeType` - If the MIME type format is invalid
+/// * `MimeTypeMismatch` - If the claimed MIME type doesn't match the detected format
+///
+/// # Security
+/// This function protects against MIME type confusion attacks by verifying that the
+/// claimed MIME type matches the actual file content. Applications cannot lie about
+/// the file type.
+pub(crate) fn validate_mime_type_matches_data(
+    data: &[u8],
+    claimed_mime_type: &str,
+) -> Result<String, ImageProcessingError> {
+    // First, validate and canonicalize the claimed MIME type
+    let canonical_claimed = validate_mime_type(claimed_mime_type)?;
+
+    // Detect the actual MIME type from the file data
+    let detected_mime_type = detect_mime_type_from_data(data)?;
+
+    // Compare the claimed type with the detected type
+    if canonical_claimed != detected_mime_type {
+        return Err(ImageProcessingError::MimeTypeMismatch {
+            claimed: canonical_claimed,
+            detected: detected_mime_type,
+        });
+    }
+
+    Ok(canonical_claimed)
+}
+
+/// Validate filename length and content
+#[allow(unused)]
```

Comment:
Instead of this maybe it should be gated to the feature (`#[cfg(feature = "mip04")]`) ?